### ISSUE TITLE
Enable using GPU for docker containers

### ISSUE THREAD

**Issue by tadejsv** [jina-ai/jina#1671]
I would like to be able to start GPU-enabled docker containers with Jina.

The following arguments need to be passed to Docker Python SDK run command for this to happen:
- `runtime`: for GPU to work, users should set this to `"nvidia"`
- `environment`: If working on a multi-GPU machine, users need to pass env variables like `CUDA_VISIBLE_DEVICES` to select which GPU the container should use
- `ipc_mode` or `shm_size`: This one is not necessary, but would be nice to have. This increases the amount of memory (RAM) available to the container - for larger models it might be needed.

Adding this would require changes in `/jina/peapods/runtimes/container/__init__.py`. If you wish, I can prepare a PR for this myself

**Comment by JoanFM** [comment]
Hey @tadej-redstone ,

This is definitely an interesting point to look at. We will take a look at it, but maybe we need a way to pass arbitraty extra parameters or extra "kwargs" to the docker sdk to be able to respond to custom needs.

**Comment by JoanFM** [comment]
> Hey @tadej-redstone ,
> 
> This is definitely an interesting point to look at. We will take a look at it, but maybe we need a way to pass arbitraty extra parameters or extra "kwargs" to the docker sdk to be able to respond to custom needs.

This would also involve changes in the parser. You are of course more than welcome to propose a PR for this solution

**Comment by tadejsv** [comment]
> > Hey @tadej-redstone ,
> > This is definitely an interesting point to look at. We will take a look at it, but maybe we need a way to pass arbitraty extra parameters or extra "kwargs" to the docker sdk to be able to respond to custom needs.
> 
> This would also involve changes in the parser. You are of course more than welcome to propose a PR for this issue

I agree, probably being able to pass custom kwargs is the best option. If this is not very hard to do, I might make a shot at a PR for it

**Comment by JoanFM** [comment]
> > > Hey @tadej-redstone ,
> > > This is definitely an interesting point to look at. We will take a look at it, but maybe we need a way to pass arbitraty extra parameters or extra "kwargs" to the docker sdk to be able to respond to custom needs.
> > 
> > 
> > This would also involve changes in the parser. You are of course more than welcome to propose a PR for this issue
> 
> I agree, probably being able to pass custom kwargs is the best option. If this is not very hard to do, I might make a shot at a PR for it

I guess we will need to check what the docker sdk allows to.

**Comment by tadejsv** [comment]
Would jina need to check if argument names for docker `run` command are ok, or would it just pass the remaining kwargs in docker, letting it raise an error if something is not right?

**Comment by JoanFM** [comment]
I think that the most maintainable way is just to pass it to docker

**Comment by tadejsv** [comment]
Alright, I think I know how this should be done then

**Comment by tadejsv** [comment]
I checked now and this does indeed seem much more complicated than I thought... It's not very clear where this should be added, and how the interactions between `argparse` and `jaml` work in Jina.

Still, I think this is an important feature - GPU applications are usually run in containers. I think the best option would be to add an optional `docker_kwargs` argument.

**Comment by JoanFM** [comment]
> I checked now and this does indeed seem much more complicated than I thought... It's not very clear where this should be added, and how the interactions between `argparse` and `jaml` work in Jina.
> 
> Still, I think this is an important feature - GPU applications are usually run in containers. I think the best option would be to add an optional `docker_kwargs` argument.

I agree, I think this should be added to the ContainerRuntime parser.

We will try to look at it. Have u been able to investigate if docker sdk has some parameter to pass extra arguments as if they were passed to a docker run command?

**Comment by tadejsv** [comment]
Well, Docker SDK by itself allows you to pass any valid docker arguments, see the documentation of the `run()` command [here]((https://docker-py.readthedocs.io/en/stable/containers.html#docker.models.containers.ContainerCollection.run)). So here's how the code in [`peapods/runtimes/container/__init__.py`](https://github.com/jina-ai/jina/blob/7e24d127f2dd26fe65e71314beb8a868e6a8a21b/jina/peapods/runtimes/container/__init__.py#L119-L127) could be extended:

``` python
        self._container = client.containers.run(uses_img,
                                                _args,
                                                detach=True,
                                                auto_remove=True,
                                                ports=ports,
                                                name=slugify(self.name),
                                                volumes=_volumes,
                                                network_mode=self._net_mode,
                                                entrypoint=self.args.entrypoint,
+                                              **docker_kwargs)
```

Before that, some basic processing of `docker_kwargs` could be implemented (purging out arguments set by jina - such as `detach` and `network_mode`), and maybe merging some others (like `ports` - in case container needs a port to connect to an external service) - but in general, once the parser gets `docker_kwargs` this would be super simple.

**Comment by JoanFM** [comment]
> Well, Docker SDK by itself allows you to pass any valid docker arguments, see the documentation of the `run()` command here. So here's how the code in [`peapods/runtimes/container/__init__.py`](https://github.com/jina-ai/jina/blob/7e24d127f2dd26fe65e71314beb8a868e6a8a21b/jina/peapods/runtimes/container/__init__.py#L119-L127) could be extended:
> 
> ```python
>         self._container = client.containers.run(uses_img,
>                                                 _args,
>                                                 detach=True,
>                                                 auto_remove=True,
>                                                 ports=ports,
>                                                 name=slugify(self.name),
>                                                 volumes=_volumes,
>                                                 network_mode=self._net_mode,
>                                                 entrypoint=self.args.entrypoint,
> +                                              **docker_kwargs)
> ```
> 
> Before that, some basic processing of `docker_kwargs` could be implemented (purging out arguments set by jina - such as `detach` and `network_mode`), and maybe merging some others (like `ports` - in case container needs a port to connect to an external service) - but in general, once the parser gets `docker_kwargs` this would be super simple.

that was the idea I had in mind.

**Comment by tadejsv** [comment]
Closed in #1690
